### PR TITLE
prov/efa: protocol v4 doc

### PR DESCRIPTION
**Pull request by wzamazon**
This PR contains document for EFA RDM protocol v4, and 7 commits to adjust naming of the implementation to make it conform to the document

### CODE REVIEW

**Comment by wzamazon** [comment]
@j-xiong Is it OK to back port this change to v1.13.x?

**Comment by j-xiong** [comment]
@wzamazon I think it's fine.

**Comment by wzamazon** [comment]
bot:aws:retest

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 1]
```diff
@@ -0,0 +1,1307 @@
...
+
+Currently, the raw address of EFA is 32 bytes, but it can be expanded in the future without
+breaking backward compatibility.
+
+## 2. Handshake sub-protocol
+
+Handshake sub-protocol serves two purposes in protocol v4.
+
+First, it is used to exchange two endpoints' capability information, which allows to introduce
+changes to protocol v4 without break backward compatibility. (section 2.1)
```
Comment (mjkoop) [inline prov/efa/docs/efa_rdm_protocol_v4.md:246]: break -> breaking

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 2]
```diff
@@ -0,0 +1,1307 @@
...
+In principal, medium message sub-protocol can be used on messages of any size. However, it is not
+recommended to use medium message sub-protocol for long messages, because it does not have flow
+control thus can overwhelm the receiver and cause network congestion. The exact size boundary for
+medium message protocol to be used is up to the implementation to decide.
+
+The following diagram illustrates its workflow:
+
+![medium message](message_medium.png)
+
+Table 3.4 describe the binary structure of a MEDIUM_RTM packet's mandatory header:
```
Comment (mjkoop) [inline prov/efa/docs/efa_rdm_protocol_v4.md:612]: This heading is repeated below

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 3]
```diff
@@ -0,0 +1,1307 @@
...
+
+#### emulated atomic protocols
+
+This section describes the 3 emulated atomic protocols: emulated write atomic, emulate fetch atomic and emulated compare atomic.
+
+The workflow of emulated write atomic is illustrated in the following diagram:
+
+![atomic_write](atomic_write.png)
+
+It is similar to emulated eager write sub-protol, expect an WRITE_RTA packet was
```
Comment (mjkoop) [inline prov/efa/docs/efa_rdm_protocol_v4.md:960]: protol -> protocol

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 4]
```diff
@@ -0,0 +1,1307 @@
...
+the requester is treating this communication as a send operation internally, and this communication is subject
+to same flow control as a long-CTS message communication does.
+
+#### emulated read features/sub-protocols
+
+This section describes two emulated read sub-protocols: emulated short read and emulated long-CTS read. Both
+sub-protocols use send/receive to emulate read. The interesting part is, in an emulated read communication,
+the responder is the sender and the requester is the receiver.
+
+The workflow of emulated short read protocol is illustrated in the following diagram:
```
Comment (mjkoop) [inline prov/efa/docs/efa_rdm_protocol_v4.md:884]: I don't believe it's been explicitly noted when this is allowed or used vs long. Although information on selection is deliberately not included, it must be less than the MTU - header size.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 5]
```diff
@@ -0,0 +1,1307 @@
...
+The field `msg_id` is provided as message ID. It is used to implement ordered atomic operations, which is supported by libfabric EFA provider,
+and is required by some application such as MPICH.
+
+The workflows of emulated fetch/compare atomic are the same, as illustrated in the following diagram:
+
+![atomic_fetch_compare](atomic_fetch_compare.png)
+
+Comparing to write atomic, the differences are:
+
+first, an FETCH_RTA/COMPARE_RTA is used to initiate the communication.
```
Comment (mjkoop) [inline prov/efa/docs/efa_rdm_protocol_v4.md:990]: first -> First

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 6]
```diff
@@ -0,0 +1,1307 @@
...
+LONGREAD_RTM is sent by the sender to initiate the communication.
+
+Like all REQ packets, a LONGREAD_RTM consists of 3 parts: mandatory header, REQ optional
+header and the application data. However, the application data part of a LONGREAD_RTM is
+special: it is not the data in the application's send buffer, but information of the
+sender buffer.
+
+In long-read message sub-protocol, sender need to construct an `read_iov`, which is
+an array of `efa_rma_iov` of application's send buffer. The `read_iov` is used
+as the appliction data in the LONGREAD_RTM packet.
```
Comment (mjkoop) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1071]: appliction -> application

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 7]
```diff
@@ -0,0 +1,1307 @@
...
+packet mandatory header is listed in table 4.3.
+
+Table: 4.3 the format of a LONGREAD_RTM packet's mandatory header
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `rma_iov_count`  | 4 | integer | `uint32_t` | message ID |
```
Comment (mjkoop) [inline prov/efa/docs/efa_rdm_protocol_v4.md:1139]: message ID is not correct here

**Review by mjkoop — COMMENTED** [review]
Just a few comments related to references to sections missing and typos

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 1 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: Thank you! fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 8]
```diff
@@ -0,0 +1,1307 @@
...
+Currently there are 4 such extra features/requests, as listed in table 2.1:
+
+Table: 2.1 a list of extra features/requests
+
+| ID | Name              |  Type    | Introduced since | Described in |
+|-|-|-|-|-|
+| 0  | RDMA read based data transfer    | extra feature | libfabric 1.10.0 | Section 4.1 |
+| 1  | delivery complete                | extra feature | libfabric 1.12.0 | Section 4.2 |
+| 2  | keep packet header length constant | extra request | libfabric 1.13.0 | Section 4.3 |
+| 3  | sender connection id in packet header  | extra request | libfabric 1.14.0 | Section 4.4 |
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: Thank you! removed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 9]
```diff
@@ -0,0 +1,1307 @@
...
+   use baseline features to carry on the communication, though the performance will
+   be sub-optimal (section 4.1).
+
+- b. the requester of the communication aborts the communication and return an error
+   to the application. For example, if application requires delivery complete, but
+   the peer does not support it (this can happen when endpoint is using libfabric 1.12,
+   but the peer is using libfabric 1.10), the requester need to return an error to the
+   application (section 4.2)
+
+Regarding extra request, if an endpoint can support an extra request the peer has, it
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 10]
```diff
@@ -0,0 +1,1307 @@
...
+should comply the request. Otherwise, it can ignore the request. Peer should be do
+one of the following:
+
+- a. carry on the communication without using the extra request. (see section 4.4
+     for example)
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: removed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 11]
```diff
@@ -0,0 +1,1307 @@
...
+
+The first 4 bytes (3 fields: `type`, `version`, `flags`) is the EFA RDM base header (section 1.3).
+
+Immediately after the base header, there are 2 fields `nextra_p3` and `extra_info`.
+
+The field `extra_info` is an array of 8 byte integers, which stores the capability of an endpoint.
+
+As mentioned before, each extra feature/request was assigned an ID when it was introduced to protocol v4.
+When constructing the handshake packet, for each extra feature it supports (or an extra request it want to impose),
+an endpoint need to toggle on a corresponding bit in the `` array. Specifically, if an endpoint supports
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 12]
```diff
@@ -0,0 +1,1307 @@
...
+
+When implementing the handshake sub-protocol, keep in mind that the application
+does not know the existence of a HANDSHAKE packet, therefore will not wait
+for its completion.
+
+For example, it is normal that a HANDSHAKE packet encounter an send error
+because peer has already been closed, because application might just send
+1 message and close the endpoint.
+
+It is also possible to close an endpoint when there are on-the-fly HANDSHAKE
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 13]
```diff
@@ -0,0 +1,1307 @@
...
+
+**REQ mandatory header** is unique for each individual REQ packet type. However, they all must start with
+the same 4 bytes EFA RDM base header (section 1.3). Recall that a base header is consisted with 3 fields:
+`type`, `version` and `flags`. Among them, `flags` warrants more discussion here, as all REQ packets share
+the same set of flags, which is listed in table 3.1:
+
+Table: 3.1 a list of REQ packet flags
+
+| Bit Id | Value | Name | meaning |
+|-|-|-|
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: Yes, fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md [hunk 14]
```diff
@@ -0,0 +1,1307 @@
...
+|  0     | 0x1    | REQ_OPT_RAW_ADDR_HDR | This REQ packet has the optional raw address header |
+|  1     | 0x2    | REQ_OPT_CQ_DATA_HDR  | This REQ packet has the optional CQ data header |
+|  2     | 0x4    | REQ_MSG              | This REQ packet is used by two-sided communication |
+|  3     | 0x8    | REQ_TAGGED           | This REQ packet is used by tagged two-sided communication |
+|  4     | 0x10   | REQ_RMA              | This REQ packet is used by an emulated RMA (read or write) communication |
+|  5     | 0x20   | REQ_ATOMIC           | This REQ packet is used by an emulated atomic (write,fetch or compare) communication |
+
+**REQ optional headers** contain additional information needed by the receiver of the REQ packets.
+As mentioned earlier, the existence of optional header in a REQ packet is indicated by bits in the `flags`
+field of the base header. There are currently 3 REQ optional headers defined:
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: Thank you! fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 3 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 4 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md:882]: In the new revision, I added a discussion later in the section.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 5 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: Fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 6 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: Fixed in new revision.

File: prov/efa/docs/efa_rdm_protocol_v4.md (hunk 7 above)
Comment (wzamazon) [reply prov/efa/docs/efa_rdm_protocol_v4.md]: Thank you! fixed in new revision.

**Review by mjkoop — APPROVED** [review]
Only recommendation would be to update the commit message to allow the reader to know the code changes were done for code readability (not just to conform to the doc -- that's the side effect)

**Pull request merged**